File: cmd/chat/generate.go
package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"net"
	"time"

	"github.com/cloudwego/eino/components/model"
	"github.com/cloudwego/eino/schema"
)

var retryBaseDelay = time.Second

func generate(ctx context.Context, llm model.ToolCallingChatModel, in []*schema.Message) (*schema.Message, error) {
	start := time.Now()
	out, err := llm.Generate(ctx, in)
	duration := time.Since(start)

	if err != nil {
		log.Printf("generate failed after %v: %v", duration, err)
		return nil, fmt.Errorf("llm generate failed: %w", err)
	}

	log.Printf("generate completed in %v", duration)
	return out, nil
}

func generateWithRetry(ctx context.Context, llm model.ToolCallingChatModel, in []*schema.Message) (*schema.Message, error) {
	const maxRetries = 3

	for attempt := 0; attempt < maxRetries; attempt++ {
		out, err := llm.Generate(ctx, in)
		if err == nil {
			if attempt > 0 {
				log.Printf("generate succeeded on retry attempt %d", attempt+1)
			}
			return out, nil
		}

		if isRetryableError(err) && attempt < maxRetries-1 {
			delay := retryBaseDelay * time.Duration(attempt+1)
			log.Printf("generate attempt %d failed: %v, retrying in %v", attempt+1, err, delay)
			select {
			case <-time.After(delay):
			case <-ctx.Done():
				return nil, fmt.Errorf("generate retry canceled: %w", ctx.Err())
			}
			continue
		}

		return nil, fmt.Errorf("after %d retries: %w", attempt+1, err)
	}

	return nil, errors.New("generateWithRetry: unexpected error")
}

func stream(ctx context.Context, llm model.ToolCallingChatModel, in []*schema.Message) (*schema.StreamReader[*schema.Message], error) {
	start := time.Now()
	stream, err := llm.Stream(ctx, in)
	duration := time.Since(start)

	if err != nil {
		log.Printf("stream failed after %v: %v", duration, err)
		return nil, fmt.Errorf("llm stream failed: %w", err)
	}

	log.Printf("stream started in %v", duration)
	return stream, nil
}

func isRetryableError(err error) bool {
	if err == nil {
		return false
	}

	var netErr net.Error
	if errors.As(err, &netErr) {
		return netErr.Timeout() || netErr.Temporary()
	}

	return false
}

File: python/chat/generate.py
"""Core chat generation functionality."""

import time
import logging
from typing import Optional
from asyncio import sleep
from contextlib import asynccontextmanager

from langchain_core.messages import BaseMessage, AIMessage
from langchain_core.language_models import BaseChatModel
from langchain_core.runnables import RunnableConfig
from langchain_core.outputs import ChatResult

logger = logging.getLogger(__name__)

# Retry configuration
RETRY_BASE_DELAY = 1.0  # seconds
MAX_RETRIES = 3


async def generate(
    model: BaseChatModel,
    messages: list[BaseMessage],
    config: Optional[RunnableConfig] = None,
) -> BaseMessage:
    """Generate a single response from the chat model."""
    start_time = time.time()

    try:
        result = await model.ainvoke(messages, config=config)
        duration = time.time() - start_time
        logger.info(f"Generate completed in {duration:.2f}s")

        # Handle different return types from the model
        if hasattr(result, 'generations') and result.generations:
            # ChatResult object
            return result.generations[0].message
        elif isinstance(result, BaseMessage):
            # Direct message object
            return result
        else:
            # Convert to AIMessage if it's just content
            return AIMessage(content=str(result))
    except Exception as e:
        duration = time.time() - start_time
        logger.error(f"Generate failed after {duration:.2f}s: {e}")
        raise


async def generate_with_retry(
    model: BaseChatModel,
    messages: list[BaseMessage],
    config: Optional[RunnableConfig] = None,
) -> BaseMessage:
    """Generate a response with retry logic for network errors."""

    for attempt in range(MAX_RETRIES):
        try:
            result = await generate(model, messages, config)
            if attempt > 0:
                logger.info(f"Generate succeeded on retry attempt {attempt + 1}")
            return result
        except Exception as e:
            if is_retryable_error(e) and attempt < MAX_RETRIES - 1:
                delay = RETRY_BASE_DELAY * (attempt + 1)
                logger.warning(f"Generate attempt {attempt + 1} failed: {e}, retrying in {delay}s")
                await sleep(delay)
                continue
            raise

    raise RuntimeError("generate_with_retry: unexpected error after all retries")


def is_retryable_error(error: Exception) -> bool:
    """Check if an error is retryable."""
    if error is None:
        return False

    # Check for timeout errors
    if "timeout" in str(error).lower():
        return True

    # Check for connection errors
    if "connection" in str(error).lower() or "network" in str(error).lower():
        return True

    # Check for rate limit errors
    if "rate limit" in str(error).lower() or "429" in str(error):
        return True

    return False